rust/trace/src/lib.rs

```diff
@@ -0,0 +1,691 @@
+            tid: sys_tid::current_tid().unwrap(),
+            pid: sys_pid::current_pid(),
+        }
+    }
+
+    fn new_instant<S>(name: S, categories: CategoriesT,
+                      payload: Option<TracePayloadT>) -> Self
+        where S: Into<StrCow>
+    {
+        let now = time::precise_time_ns();
+        Self {
+            sample_id: SAMPLE_COUNTER.fetch_add(1, AtomicOrdering::Relaxed),
+            name: name.into(),
+            categories: categories,
+            start_ns: now,
+            payload: payload,
+            end_ns: now,
+            sample_type: SampleType::Instant,
+            tid: sys_tid::current_tid().unwrap(),
+            pid: sys_pid::current_pid(),
+        }
+    }
+}
+
+impl PartialEq for Sample {
+    fn eq(&self, other: &Sample) -> bool {
+        self.sample_id == other.sample_id
+    }
+}
+
+impl Eq for Sample {}
+
+impl PartialOrd for Sample {
+    fn partial_cmp(&self, other: &Sample) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Sample {
+    fn cmp(&self, other: &Sample) -> cmp::Ordering {
+        self.sample_id.cmp(&other.sample_id)
+    }
+}
+
+impl Hash for Sample {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.sample_id.hash(state);
+    }
+}
+
+pub struct SampleGuard {
+    sample: Option<Sample>,
+}
+
+impl SampleGuard {
+    #[inline]
+    fn new_disabled() -> Self {
+        Self {
+            sample: None
+        }
+    }
+
+    #[inline]
+    fn new<S>(name: S, categories: CategoriesT, payload: Option<TracePayloadT>)
+        -> Self
+        where S: Into<StrCow>
+    {
+        Self {
+            sample: Some(Sample::new(name, categories, payload))
+        }
+    }
+}
+
+impl Drop for SampleGuard {
+    fn drop(&mut self) {
+        if let Some(ref mut sample) = self.sample {
+            sample.end_ns = time::precise_time_ns();
+            record_sample(sample);
+        }
+    }
+}
+
+struct Trace {
+    enabled: AtomicBool,
+    samples: Mutex<FixedLifoDeque<Sample>>,
+}
+
+impl Trace {
+    fn new() -> Self {
+        Self {
+            enabled: AtomicBool::new(false),
+            samples: Mutex::new(FixedLifoDeque::new())
+        }
+    }
+}
+
+lazy_static! { static ref TRACE : Trace = Trace::new(); }
+
+pub fn enable_tracing() {
+    enable_tracing_with_config(&Config::default());
+}
+
+pub fn enable_tracing_with_config(config: &Config) {
+    let mut all_samples = TRACE.samples.lock().unwrap();
+    all_samples.reset_limit(config.max_samples());
+    TRACE.enabled.store(true, AtomicOrdering::Relaxed);
+}
+
+pub fn disable_tracing() {
+    let mut all_samples = TRACE.samples.lock().unwrap();
+    all_samples.reset_limit(0);
+    SAMPLE_COUNTER.store(0, AtomicOrdering::Relaxed);
+    TRACE.enabled.store(false, AtomicOrdering::Relaxed);
+}
+
+#[inline]
+fn is_enabled() -> bool {
+    TRACE.enabled.load(AtomicOrdering::Relaxed)
+}
+
+pub fn trace<S>(name: S, categories: CategoriesT)
```

Comment:
This looks much cleaner, I presume you ran rustfmt over it? At some point we should probably have some common formatting guidelines.